### PR TITLE
Use AfterReturningAdvice Instead of PostSubmission

### PR DESCRIPTION
**Pull request by sherrif10**
Am facing a null pointer exception when i try to save an encounter symptom form cc @dkayiwa

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/dss/api/action/SymptomFormAction.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                    }
+                    capturedObsList.add(obs); // Add the observed value to the list of captured observations
+                });
+    
+                Encounter newEncounter = new Encounter();
+                newEncounter.setPatient(symptomEncounter.getPatient());
+                newEncounter.setLocation(symptomEncounter.getLocation());
+                newEncounter.setEncounterProviders(symptomEncounter.getActiveEncounterProviders());
+                newEncounter.setEncounterDatetime(symptomEncounter.getEncounterDatetime());
+                newEncounter.setEncounterType(symptomEncounter.getEncounterType());
```
Comment (sherrif10) [inline api/src/main/java/org/openmrs/module/dss/api/action/SymptomFormAction.java]: Hello @mozzy. Arent we supposed to create a new encounter here ?

Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/dss/api/action/SymptomFormAction.java]: i removed creation of new encounters

File: api/src/main/java/org/openmrs/module/dss/api/action/SymptomFormAction.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        if (symptomEncounter.getObs() != null) {
+            Optional<Obs> destinationObs = symptomEncounter.getObs().stream()
+                    .filter(obs -> obs.getConcept().getConceptId().equals(DESTINATION_CONCEPT_ID)).findFirst();
+            if (destinationObs.isPresent()) {
+                if (destinationObs.get().getValueText().equals(AIMODEL_DESTINATION)) {
+                    Set<Concept> testConcepts = new HashSet<>();
+                    List<Obs> capturedObsList = new ArrayList<>();
+    
+                    symptomEncounter.getObs().forEach(obs -> {
+                        if (obs.getValueCoded() != null) {
```
Comment (mozzy11) [inline api/src/main/java/org/openmrs/module/dss/api/action/SymptomFormAction.java]: What are trying to do here ??

Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/dss/api/action/SymptomFormAction.java]: Thanks for the review, Basing on the last chat, you had advised me to use AOP for after returning advice.  Here is the current Pr

Comment (mozzy11) [reply api/src/main/java/org/openmrs/module/dss/api/action/SymptomFormAction.java]: i see. 
the you should remove the action class

Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/dss/api/action/SymptomFormAction.java]: Yes i updated the code.

File: api/src/main/java/org/openmrs/module/dss/api/Advice/EncounterServiceAfterAdvice.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.lang.reflect.Method;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.openmrs.Encounter;
+import org.openmrs.api.EncounterService;
+import org.springframework.aop.AfterReturningAdvice;
+import java.util.logging.Logger;
+
+public class EncounterServiceAfterAdvice implements AfterReturningAdvice {
```
Comment (mozzy11) [inline api/src/main/java/org/openmrs/module/dss/api/Advice/EncounterServiceAfterAdvice.java:16]: you dont need this class.
you already have another AOP class

File: api/src/main/java/org/openmrs/module/dss/api/aspect/LoggingAspect.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package org.openmrs.module.dss.api.aspect;
+
+import org.springframework.context.annotation.EnableAspectJAutoProxy;
+import org.springframework.stereotype.Component;
+
+@EnableAspectJAutoProxy
```
Comment (mozzy11) [inline api/src/main/java/org/openmrs/module/dss/api/aspect/LoggingAspect.java]: what was this for ???

File: omod/src/main/resources/webModuleApplicationContext.xml [hunk 5]
```diff
@@ -25,4 +25,12 @@
 	<bean class="org.openmrs.ui.framework.StandardModuleUiConfiguration">
 		<property name="moduleId" value="${project.parent.artifactId}"/>
 	</bean>
+	<bean id="userContextProxy" class="org.springframework.aop.framework.ProxyFactoryBean">
+    <property name="target" ref="context" />
+    <property name="interceptorNames">
+        <list>
+            <value>EncounterServiceAfterAdvice</value>
```
Comment (mozzy11) [inline omod/src/main/resources/webModuleApplicationContext.xml]: what was this for ??

File: omod/src/main/resources/config.xml [hunk 6]
```diff
@@ -107,7 +110,10 @@
 	-->
 	<!-- /Servlets -->
 	
-	
+	<advice>
+     <point>org.openmrs.api.EncounterService</point>
+     <class>org.openmrs.module.dss.api.aspect.EncounterServiceAfterAdvice</class>
+     </advice>
 	<!-- Internationalization -->
```
Comment (mozzy11) [inline omod/src/main/resources/config.xml:117]: i gues this is the only config you need to register your AOP advise

File: omod/src/main/resources/webModuleApplicationContext.xml (hunk 5 above)
Comment (sherrif10) [reply omod/src/main/resources/webModuleApplicationContext.xml]: Here i was injecting this class as a proxy.

File: api/src/main/java/org/openmrs/module/dss/api/aspect/LoggingAspect.java (hunk 4 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/dss/api/aspect/LoggingAspect.java]: Resolved

File: omod/target/classes/webModuleApplicationContext.xml [hunk 7]
```diff
@@ -25,4 +25,12 @@
 	<bean class="org.openmrs.ui.framework.StandardModuleUiConfiguration">
 		<property name="moduleId" value="dss"/>
 	</bean>
+	<bean id="userContextProxy" class="org.springframework.aop.framework.ProxyFactoryBean">
+    <property name="target" ref="context" />
+    <property name="interceptorNames">
+        <list>
+            <value>EncounterServiceAfterAdvice</value>
```
Comment (mozzy11) [inline omod/target/classes/webModuleApplicationContext.xml:32]: do you need to register this here ??

Comment (sherrif10) [reply omod/target/classes/webModuleApplicationContext.xml:32]: Its not part of this class. it was not supposed to be committed.

File: api/src/main/java/org/openmrs/module/dss/api/Advice/EncounterServiceAfterAdvice.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  Logger.getAnonymousLogger().info("======== in After Advice " + method.getName() + "========");
+
+  if (method.getName().equals("saveEncounter") && args[0] != null && returnValue != null) {
+    Encounter encounter = (Encounter) returnValue;
+      if (encounter != null && encounter.getObs() != null) {
+        if (encounterService != null) {
+          Set<Obs> observations = encounter.getObs();
+          // Iterate through the observations and add them to the ArrayList
+          ArrayList<Obs> observationList = new ArrayList<>();
+          for (Obs obs : observations) {
```
Comment (sherrif10) [inline api/src/main/java/org/openmrs/module/dss/api/Advice/EncounterServiceAfterAdvice.java]: Hello @mozzy11,  Thanks for this, Hope this is good to go, Am able to see concepts being returned in the array List, My other task to create a http client and post those concept ids on to the server

**Comment by sherrif10** [comment]
Thanks @mozzy11  @dkayiwa for guidance.

**Pull request merged**